RFC: Negation types (#29)
luau-lang/rfcs: docs/negation-types.md

# Negation types

## Summary

A type that represents the complement of the type being negated, which is a union type of all inhabited types (`unknown`) excluding the type being negated.

## Motivation

> Type refinements will produce negation types to get the complementation. For the most part, users will never need to write these negation types themselves, except for when they want to _maintain_ some invariances beyond the scope of the branch.

Reviewer: ```suggestion
Type refinements will produce negation types to get the complementation. For the most part, users will never need to write these negation types themselves, except for when they want to _maintain_ some invariants beyond the scope of the branch.
```

> A [cofinite set](https://en.wikipedia.org/wiki/Cofiniteness) arises when you have a negation type in conjunction with intersection types. For example, `string & ~"a"` is a cofinite string set that accepts any `string` excluding `"a"`. This typically happens all the time with type refinements, where `typeof(x) == "string" and x ~= "a"` gives rise to the type `string & ~"a"`.

Reviewer: ```suggestion
A [cofinite set](https://en.wikipedia.org/wiki/Cofiniteness) arises when you have a negation type in conjunction with intersection types. For example, `string & ~"a"` is a cofinite string set that accepts any `string` excluding `"a"`. This happens often with type refinements where `typeof(x) == "string" and x ~= "a"` gives rise to the type `string & ~"a"`.
```

## Design

### Syntax

The notation is `~T` where `T` is a type annotation. It binds tightly to the type it negates, so `~T | U` is `(~T) | U`, not `~(T | U)`. There's really not much to it, syntax-wise. It has the same precedence as the `not` expression, which is why `~T | U` is `(~T) | U` in the same way `not t or u` is `(not t) or u`.

We propose to allow users to write such a type. Formally:

```ebnf
ty ::= `~` ty

>      | name [`<` [ty] {`,` ty} `>`]

Author: ```suggestion
     | name [`<` [ty {`,` ty}] `>`]
```

1. `~~any`
2. `~~(unknown | *error*)`
3. `~(~unknown & ~*error*)`
4. `~(never & unknown)`
5. `~never`
6. `unknown` (not equivalent to `any`!)

As you can see, since through the series of rewrites `~~any` did not produce a type equivalent to `any`, our basis set must be `unknown`, and negation of an error suppression is still an error suppressing type. Therefore `~*error*` is `*error*`. This fixes the double-negation inconsistency:

1. `~~any`
2. `~~(unknown | *error*)`
3. `~(~unknown & ~*error*)`
4. `~(never & *error*)`
5. `~never | ~*error*`
6. `unknown | *error*` (`any` is an alias to `unknown | *error*`, so equivalence is achieved)

### Restrictions

We currently only support negation of _testable_ types. Intuitively, a testable type is one that can be tested by type refinements at runtime. This includes singleton types, primitive types, and classes. Of course, unions and intersections can be negated as well if and only if all of its constituent types are also testable types, due to the distributivity property of unions and intersections.

> Since types like `{ p: P }` and `(T) -> U` are structural types and the runtime offers no way to test for them, they cannot be negated. An attempt to negate this will just turn the negation type into an error suppressed type and also a type error. This means `~{ p: P }` and `~(T) -> U` are nonsensical.

Reviewer: We should probably explicitly error if a user writes an annotation that is the negation of a structural type.

Author: There was "and also a type error" but I can extract that into a separate sentence in this paragraph.

Author: Word ordering is super important and it's rather annoying. Function type. Type function. Refinement type. Type refinement.

Another restriction is that negation types cannot negate generic types. So `<T>(T) -> ~T` is also not legible.

### Implementation

Most of the implementation of negation types are already in place. The three things missing are:

1. the syntax,
2. some guarantee that no negation types survive if it negates non-testables, and
3. a type error if it negates a non-testable type in any way, shape, or form.

The parser change for this is trivial, so this is of no concern.

We can use type families to have this guarantee. Add a `negate<T>` type family which would be internal to the type inference engine, and have the syntax `~T` produce that type family, not a `NegationType`.

> As for the type error, we can just consider `negate<{ p: P }>` to be an uninhabited type family, and resolve that as an error type.

Reviewer: I think this is very sensible, and aligned with where future proposals will go too (e.g. type indexing syntax like `SomeTableType.SomeProperty` will probably want to resolve to `index<SomeTableType, SomeProperty>`). Seems like the right call to me.
